relay-server/src/actors/relays.rs

```diff
@@ -79,49 +147,37 @@ impl RelayState {
     }
 }
 
-#[derive(Debug)]
-struct RelayInfoChannel {
-    sender: oneshot::Sender<Option<RelayInfo>>,
-    receiver: Shared<oneshot::Receiver<Option<RelayInfo>>>,
-}
-
-impl RelayInfoChannel {
-    pub fn new() -> Self {
-        let (sender, receiver) = oneshot::channel();
-        RelayInfoChannel {
-            sender,
-            receiver: receiver.shared(),
-        }
-    }
+type FetchResult = Result<GetRelaysResult, HashMap<RelayId, Vec<Sender<GetRelayResult>>>>;
 
-    pub fn send(self, value: Option<RelayInfo>) -> Result<(), Option<RelayInfo>> {
-        self.sender.send(value)
-    }
-
-    pub fn receiver(&self) -> Shared<oneshot::Receiver<Option<RelayInfo>>> {
-        self.receiver.clone()
-    }
-}
-
-pub struct RelayCache {
+#[derive(Debug)]
+pub struct RelayCacheService {
+    static_relays: HashMap<RelayId, RelayInfo>,
+    relays: HashMap<RelayId, RelayState>,
+    senders: HashMap<RelayId, Vec<Sender<GetRelayResult>>>,
```

Comment:
This is a substantial change. Instead of spawning a long-running future for every request call, we simply queue up the incoming message's `Sender`.   If the relay info is already cached, this is skipped, of course.

This requires us to allocate the more incoming requests are waiting for a result, although it should still cause less overhead than maintaining a tokio task each. While queued, these senders do not consume CPU.